tools/rapidres.1.md
% RAPIDRES(1) Rapicorn-@BUILDID@ | Rapicorn Manual
%
% @FILE_REVISION@


# NAME
rapidres - A Rapicorn binary resource file to C++ code converter


# SYNOPSIS
**rapidres** [**-h**] [**-v**] [*files*...]


# DESCRIPTION

Rapidres reads binary files and generates C++ source code to compile the
binary contents into a C++ program. The generated source code relies on
the definition of the following two macros:

**RAPICORN_RES_STATIC_DATA**(*symbol*)
:   Defines a static variable to hold a char array. By default, the data
    is read out of an input file, compressed (with zlib) and written out
    in C string syntax.

**RAPICORN_RES_STATIC_ENTRY**(*symbol*, *path*, *size*)
:   Defines a static variable with metadata about *symbol*.
    The *symbol* is the same as passed into **RAPICORN_RES_STATIC_DATA()**,
    the *path* designates the filename of the input file to identify the
    binary contents at runtime. The *size* argument corresponds to the
    actual uncompressed size of the data stored
    with **RAPICORN_RES_STATIC_DATA()**.


# OPTIONS

**Rapidres** follows the usual GNU command line syntax, with long options starting with two dashes ('-').

**-h**, **\--help**
:   Print usage information.

**-v**, **\--version**
:   Print version and file paths.


# EXAMPLES

Generate source code to compile a simple file into an executable:

**echo** Hello World > hello.txt	\
**rapidres** hello.txt

This generates something along the lines of:

    /* rapidres file dump of hello.txt */
    RAPICORN_RES_STATIC_DATA (hello_txt) = "Hello World\n"; // 12 + 1
    RAPICORN_RES_STATIC_ENTRY (hello_txt, "hello.txt", 12);


# AUTHORS
Tim Janik.


# SEE ALSO

**rapidrun**(1), [**Rapicorn Website**](https://rapicorn.testbit.org)


NEWS
Rapicorn 17.0.0~rc1:

* Support clang++-3.6.2.
* Fixed uild on Arch linux [milkmiruku]
* Fixed build on Kubuntu 16.04 [stw]
* Fixed bad_weak_ptr exceptions.
* Added support g_main_context() integration.
* Improved the quality of rendering stretched SVG files.
* Fixed spacing calculcations, text block creation and resizing priorities.
* Parse .ini files from $XDG_CONFIG_HOME, $XDG_CONFIG_DIRS and /etc/rapicorn/.
* Support selectable container children.
* Support Capturing event phase.
* Automatically allow shrinking for all spreading widgets to preserve others.
* Fixed focus movemenet to (partially) obscured widgets.
* Use viewport window relative coordinates for all widgets.
* Simplified code in several places.
* Improved rendering speed significantly.
* Introduced SelectableItem for lists and remove some dated code.
* Cleaned up the ancestry cache handling and allow Window.destroy().
* Simplified ThemeInfo and styling.
* Added support for SVGs in the widget factory.
* Use 16 bit for widget states and add TOGGLED and STASHED.
* Fix the SVG ink rectangle where libsvg only exports ints instead of doubles.
* Add ElementPainter: with 'element' and 'filler' properties to render SVGs.
* Allow normal buttons to toggle.
* Added ShapePainter that can render circle, diamond, square and star.
* Support a Button.label property.
* Add a RadioButton implementation.
* Improved filewview UI.
* SVG theme updates.
* Use <Element id="some-identifier" declare="new-element"/> in XML UIs.
* Added Widget properties acceleratable, hover, panel, receives_default.
* Also add Widget properties selected, insensitive and active.
* Consistently use int for size calculations everywhere.
* Optimized resizing and invalidation region calculations.
* Cleaned up the code and build rules in several places.
* Support fine grained size negotiations.
* Support per widget offline surfaces.
* Better documented ABI versioning affects.
* Improved documentation.
* Added test categorization and numbering.
* Build Rapicorn via non-recursive sub Makefiles.
* Fixed use-after-free in DisplayWindowX11.
* Avoid overlinking in rapicorn.pc.
* Require srcdir=destdir upon distcheck.
* Added backtrace generator via addr2line.
* Added support for notifications from Aida remote calls.
* Added support for event loop source groups.
* Majorly improved Docker build setup.

Rapicorn 16.0.0:	(2016-04-07)

* Remove internal librsvg copy, require an installed librsvg-2.0.
* Provide enum meta information for enum values stored in Any.
* Provide change notification signals for Parameter.
* Cleanup threads via atexit(3).
* Added support for auxillary data on IDL sequence and record definitions.
* Added compatibility code to downgrade cython requirement to 0.20.1 for trusty.
* Added fixes for g++-4.8.4 and libtool >= 1.5.2.
* Grow Any content to enable structured type and small string optimizations.
* Remove remove Atomic<> in favour of the "Construct On First Use Idiom".
* Improved exception specifications to satisfy cython-0.23.3.
* Reduced unit testing time, individually and by tuning the benchmark timers.
* Use xvfb-run(1) to run X11 related unit tests.
* Support and use the TAP protocol for unit test evaluation.
* New versioning: Releases increase MAJOR for major API changes.
* New versioning: Releases increase MINOR in case of any ABI breaks.
* New versioning: Releases increase MICRO for backwards compatibility.
* New versioning: The MAJOR number ideally matches the release year.
* The Python module alias remans 'Rapicorn', stored in Rapicorn_15/.
* C++ namespace changed to 'Rapicorn', files install under 'rapicorn-$MAJOR'.
* The librapicorn shared library supports installation of multiple versions.
* Many build system fixes and integration of travis-ci builds.
* Automatically built Debian packages are now on http://bintray.com/beast-team/.
* The file debian/copyright now contains an accurate licensing account.
* Install cython headers to allow IDL Python bindings outside Rapicorn.
* Add AutoSeeder class to provide good non-deterministic seeding entropy.
* Add PCG-32 which provides fast but well distributed pseudo random numbers.
* Include a random_secret() in hash functions to avoid collision attacks.
* Use Python enum classes and C++11 enum classes in IDL code generation.
* Fixed shebang to reliably find Python2.7. [Milk Brewster]

Rapicorn 15.09.0:	(2015-09-16)

* Added theming support and $RAPICORN_THEME to override theme selection.
* Update RSVG rendering bits.
* Added layer and state painting widgets.
* Provide rapidres(1): A binary resource file to C++ code converter.
* Added missing manual page documentation.
* Added ability to render SVG fragments.
* Added widget for state painting and for stretched SVG painting.
* Added ElementPainter to render and resize widgets from stretched SVGs.
* Added FocusPainter for focus state dependent decoration.
* Added support for attribute="@eval some_variable" syntax.
* Added example for interactive code editing and execution.
* Factory cleanups, widget have an 'id', properties have a 'name' attribute.
* Allow source priorities to affect scheduling across other event loops.
* Major improvements to the Aida Python bindings by making use of Cython.
* IDL records and sequences now have different client and server code.
* Eliminated '=0' syntax for IDL methods.
* Added C++ visitors to IDL objects, records and sequences for INI and XML.
* Improved enum introspection and IDL marshalling.
* Major improvements to type support in Any getters and setters.
* Provide forward and backward conversion between Any and records/sequences.
* Added Parameter abstraction to simlify record field and property edits.
* Display Window code cleanup and optimization.
* Automate memory bookkeeping with shared_ptr for all objects and widgets.
* Various improvements to the C++11 support, compilaion and build system.

Rapicorn 14.10.0:	(2014-10-09)

* Project is re-license under MPLv2.

tools/rapidres.cc
// This Source Code Form is licensed MPL-2.0: http://mozilla.org/MPL/2.0
#include "rcore/cxxaux.hh"
#include "config/config.h"
#include "topbuildid.hh"
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <zlib.h>
#include <vector>
#include <string>
#include <sys/types.h>
#include <regex.h>
using namespace Rapicorn;

/* rapidres - small C source compression utility
 * Implemented without linking against librapicorn, as it may be used during the rapicorn build process.
 */

typedef std::string String;

static const char *main_argv0 = "rapidres";

static void zintern_error  (const char *format, ...) RAPICORN_PRINTF (1, 2);

static void
zintern_error (const char *format, ...)
{
  char buffer[4096];
  va_list args;
  va_start (args, format);
  vsnprintf (buffer, sizeof (buffer), format, args);
  va_end (args);
  buffer[sizeof (buffer) - 1] = 0;
  fprintf (stderr, "\n%s: ERROR: %s\n", main_argv0, buffer);
  _exit (1);
}

static String strip_prefix;

typedef struct {
  uint pos;
  bool pad;
} Config;
static Config config_init = { 0, 0 };

static inline void
print_uchar (Config *config,
	     uint8 d)
{
  if (config->pos > 70)
    {
      printf ("\"\n  \"");
      config->pos = 3;
      config->pad = false;
    }
  if (d == '\\')
    {
      printf ("\\\\");
      config->pos += 2;
    }
  else if (d == '\n')
    {
      printf ("\\n");
      config->pos += 2;
    }
  else if (d == '"')
    {
      printf ("\\\"");
      config->pos += 2;
    }
  else if (d < 32 || d > 126 || d == '?')
    {
      printf ("\\%o", d);
      config->pos += 1 + 1 + (d > 7) + (d > 63);
      config->pad = d < 64;
      return;
    }
  else if (config->pad && d >= '0' && d <= '9')
    {
      printf ("\"\"");
      printf ("%c", d);
      config->pos += 3;
    }
  else
    {
      printf ("%c", d);
      config->pos += 1;
    }
  config->pad = false;
}

static void
gen_zfile (const String &file)
{
  // create nice resource path
  String fname = file;
  // strip common path prefix
  if (!strip_prefix.empty())
    {
      regex_t rx;
      if (regcomp (&rx, strip_prefix.c_str(), REG_EXTENDED | REG_NEWLINE) == 0)
        {
          regmatch_t pmatch[1];
          if (regexec (&rx, fname.c_str(), ARRAY_SIZE (pmatch), pmatch, 0) == 0 &&
              pmatch[0].rm_so == 0 && size_t (pmatch[0].rm_eo) < fname.size())
            {
              fname = fname.substr (pmatch[0].rm_eo);
            }
          regfree (&rx);
        }
    }
  // create C identifier
  String ident = fname;
  // substitute [/.-] with _
  for (size_t i = 0; i < ident.size(); i++)
    if (strchr ("./-", ident[i]))
      ident[i] = '_';
  // check identifier chars
  for (size_t i = 0; i < ident.size(); i++)
    if ((ident[i] >= 'a' && ident[i] <= 'z') ||
        (ident[i] >= 'A' && ident[i] <= 'Z') ||
        (i && ident[i] >= '0' && ident[i] <= '9') ||
        ident[i] == '_')
      ; // fine
    else
      zintern_error ("file name contains non-symbol characters: %s", file.c_str());
  // file processing
  std::vector<uint8> vdata;
  uint i, dlen = 0, mlen = 0;
  uLongf clen;
  Config config;
  FILE *f = fopen (file.c_str(), "r");
  if (!f)
    zintern_error ("failed to open \"%s\": %s", file.c_str(), strerror (errno));
  do
    {
      if (mlen <= dlen + 1024)
	{
	  mlen += 8192;
          vdata.resize (mlen);
	}
      dlen += fread (&vdata[dlen], 1, mlen - dlen, f);
    }
  while (!ferror (f) && !feof (f));

  if (ferror (f))
    zintern_error ("failed to read from \"%s\": %s", file.c_str(), strerror (errno));

  std::vector<uint8> cdata;

  int result;
  const char *err;
  clen = dlen + dlen / 100 + 64;
  cdata.resize (clen);
  result = compress2 (&cdata[0], &clen, &vdata[0], dlen, Z_BEST_COMPRESSION);
  switch (result)
    {
    case Z_OK:
      err = NULL;
      break;
    case Z_MEM_ERROR:
      err = "out of memory";
      break;
    case Z_BUF_ERROR:
      err = "insufficient buffer size";
      break;
    default:
      err = "unknown error";
      break;
    }
  if (err)
    zintern_error ("while compressing \"%s\": %s", file.c_str(), err);

  printf ("/* rapidres file dump of %s */\n", file.c_str());

  /* two things to consider for using the compressed data:
   * 1) for the reader code, pack_size + 1 must be smaller than data_size to identify compressed data.
   * 2) using compressed data requires runtime unpacking overhead and extra dynamic memory allocation,
   *    so it should provide a *significant* benefit if it's used.
   */
  const bool compress_resource = clen <= 0.75 * dlen && clen + 1 < dlen;
  const size_t rlen = compress_resource ? clen : dlen;
  const uint8 *rdata = rlen == dlen ? &vdata[0] : &cdata[0];

  config = config_init;
  printf ("RAPICORN_RES_STATIC_DATA (%s) =\n  \"", ident.c_str());
  for (i = 0; i < rlen; i++)
    print_uchar (&config, rdata[i]);
  printf ("\"; // %zu + 1\n", rlen);

  config = config_init;
  printf ("RAPICORN_RES_STATIC_ENTRY (%s, \"", ident.c_str());
  for (i = 0; i < fname.size(); i++)
    print_uchar (&config, fname[i]);
  printf ("\", %u);\n", dlen);

  fclose (f);
}

static int
help (int exitcode)
{
  printf ("usage: rapidres [-h] [-v] [-s prefix] [files...]\n");
  if (exitcode != 0)
    exit (exitcode);
  printf ("  -h, --help    Print usage information\n");
  printf ("  -v, --version Print version and file paths\n");
  printf ("  -s PREFIX     Strip PREFIX from resource names\n");
  printf ("Generate compressed C source code for each file.\n");
  exit (0);
}

int
main (int argc, char *argv[])
{
  main_argv0 = argv[0];
  std::vector<String> arg_strings;

  for (int i = 1; i < argc; i++)
    {
      if (strcmp ("-h", argv[i]) == 0 || strcmp ("--help", argv[i]) == 0)
	{
	  return help (0);
	}
      else if (strcmp ("-v", argv[i]) == 0 || strcmp ("--version", argv[i]) == 0)
	{
          printf ("rapidrun (Rapicorn utilities) %s (Build ID: %s)\n", RAPICORN_VERSION, ::Internal::buildid());
          printf ("Copyright (C) 2007 Tim Janik.\n");
          printf ("This is free software and comes with ABSOLUTELY NO WARRANTY; see\n");
          printf ("the source for copying conditions. Sources, examples and contact\n");
          printf ("information are available at https://rapicorn.testbit.org/.\n");
          exit (0);
	}
      else if (strcmp ("-s", argv[i]) == 0 && i + 1 < argc)
        {
          strip_prefix = argv[i + 1];
          i++;
        }
      else
	arg_strings.push_back (argv[i]);
    }

  if (arg_strings.size() == 0)
    return help (1);

  for (size_t i = 0; i < arg_strings.size(); i++)
    {
      const String file = arg_strings[i + 0];
      gen_zfile (file);
    }

  return 0;
}
